File: api-kt/src/main/kotlin/site/daydream/mote/model/Drive.kt
package site.daydream.mote.model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotEmpty
import jakarta.validation.constraints.Positive
import java.net.URLConnection

// ---------- DB row models ----------

data class DriveNode(
    val id: Long = 0,
    @JsonProperty("parent_id")
    val parentId: Long? = null,
    val type: String = "file", // "folder" | "file"
    val name: String = "",
    @JsonIgnore
    val blobPath: String? = null,
    val size: Long? = null,
    @JsonInclude(JsonInclude.Include.NON_NULL)
    val hash: String? = null,
    @JsonProperty("deleted_at")
    @JsonInclude(JsonInclude.Include.NON_NULL)
    val deletedAt: Long? = null,
    @JsonIgnore
    val deleteBatchId: String? = null,
    @JsonProperty("created_at")
    val createdAt: Long = 0,
    @JsonProperty("updated_at")
    val updatedAt: Long = 0,

    /** slash-joined ancestor path; populated only by global search responses */
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    val path: String = "",

    /** active share count for this file node; emitted only when > 0 */
    @JsonProperty("share_count")
    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
    val shareCount: Int = 0,
) {
    val ext: String?
        @JsonProperty("ext")
        get() = if (type != "file") null
        else {
            val n = name
            val i = n.lastIndexOf('.')
            if (i < 0) "" else n.substring(i).lowercase()
        }

    @get:JsonProperty("mime_type")
    val mimeType: String?
        get() {
            if (type != "file") return null
            return URLConnection.guessContentTypeFromName(name) ?: extMime(ext) ?: "application/octet-stream"
        }

    companion object {
        private fun extMime(e: String?): String? = when (e) {
            ".webp" -> "image/webp"
            ".svg" -> "image/svg+xml"
            ".md" -> "text/markdown"
            ".mjs" -> "application/javascript"
            ".json" -> "application/json"
            ".yaml", ".yml" -> "application/yaml"
            ".mp4" -> "video/mp4"
            ".webm" -> "video/webm"
            ".mp3" -> "audio/mpeg"
            ".wav" -> "audio/wav"
            ".ogg" -> "audio/ogg"
            ".zip" -> "application/zip"
            ".7z" -> "application/x-7z-compressed"
            ".tar" -> "application/x-tar"
            ".gz" -> "application/gzip"
            ".pdf" -> "application/pdf"
            else -> null
        }
    }
}

data class DriveUpload(
    val id: String = "",
    val parentId: Long? = null,
    val name: String = "",
    val size: Long = 0,
    val chunkSize: Long = 0,
    val totalChunks: Int = 0,
    val receivedMask: ByteArray = ByteArray(0),
    val status: String = "uploading",
    val expiresAt: Long = 0,
    val createdAt: Long = 0,
    val updatedAt: Long = 0,
)

data class DriveShare(
    val id: Long = 0,
    val nodeId: Long = 0,
    val tokenHash: String = "",
    val tokenPrefix: String = "",
    val passwordHash: String? = null,
    val expiresAt: Long? = null,
    val createdAt: Long = 0,
)

data class ShareWithNode(
    val share: DriveShare,
    val name: String,
    val size: Long,
    val parentId: Long?,
    val path: String,
)

data class DriveBreadcrumb(
    val id: Long,
    val name: String,
)

// ---------- Request DTOs ----------

data class DriveCreateFolderRequest(
    @JsonProperty("parent_id")
    val parentId: Long? = null,
    @field:NotBlank
    val name: String,
)

data class DriveRenameRequest(
    @field:Positive
    val id: Long,
    @field:NotBlank
    val name: String,
)

data class DriveMoveRequest(
    @field:NotEmpty
    val ids: List<Long>,
    @JsonProperty("new_parent_id")
    val newParentId: Long? = null,
)

data class DriveDeleteRequest(
    @field:NotEmpty
    val ids: List<Long>,
)

data class DriveRestoreRequest(
    @field:Positive
    val id: Long,
)

data class DrivePurgeRequest(
    @field:NotEmpty
    val ids: List<Long>,
)

data class DriveUploadInitRequest(
    @JsonProperty("parent_id")
    val parentId: Long? = null,
    @field:NotBlank
    val name: String,
    @field:Positive
    val size: Long,
    @JsonProperty("chunk_size")
    val chunkSize: Long = 0,
)

data class DriveUploadCompleteRequest(
    @JsonProperty("upload_id")
    @field:NotBlank
    val uploadId: String,
    @JsonProperty("on_collision")
    val onCollision: String = "ask", // ask|overwrite|rename|skip
)

data class DriveShareCreateRequest(
    @JsonProperty("node_id")
    @field:Positive
    val nodeId: Long,
    val password: String? = null,
    @JsonProperty("expires_at")
    val expiresAt: Long? = null,
)

data class DriveShareRevokeRequest(
    @JsonProperty("share_id")
    @field:Positive
    val shareId: Long,
)

// ---------- Response DTOs ----------

data class DriveUploadInitResponse(
    @JsonProperty("upload_id") val uploadId: String,
    @JsonProperty("total_chunks") val totalChunks: Int,
    @JsonProperty("chunk_size") val chunkSize: Long,
    @JsonProperty("received_chunks") val receivedChunks: List<Int>,
)

data class DriveUploadStatusResponse(
    @JsonProperty("upload_id") val uploadId: String,
    @JsonProperty("total_chunks") val totalChunks: Int,
    @JsonProperty("chunk_size") val chunkSize: Long,
    val size: Long,
    @JsonProperty("received_chunks") val receivedChunks: List<Int>,
    val status: String,
)

data class DriveShareDto(
    val id: Long,
    @JsonProperty("node_id") val nodeId: Long,
    @JsonProperty("has_password") val hasPassword: Boolean,
    @JsonProperty("expires_at") val expiresAt: Long?,
    @JsonProperty("created_at") val createdAt: Long,
    @JsonInclude(JsonInclude.Include.NON_NULL)
    val url: String? = null,
    @JsonInclude(JsonInclude.Include.NON_NULL)
    val token: String? = null,
)

data class DriveSharedItemDto(
    val id: Long,
    @JsonProperty("node_id") val nodeId: Long,
    @JsonProperty("parent_id") val parentId: Long?,
    @JsonProperty("has_password") val hasPassword: Boolean,
    @JsonProperty("expires_at") val expiresAt: Long?,
    @JsonProperty("created_at") val createdAt: Long,
    val name: String,
    val size: Long,
    val path: String,
)

File: api-rs/src/model/drive.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use std::path::Path;

/// A folder or file in the drive tree. `mime_type` and `ext` are derived from
/// `name` at serialization time and are not stored in the database (matching
/// the Go implementation).
#[derive(Debug, Clone, FromRow)]
pub struct DriveNodeRow {
    pub id: i64,
    pub parent_id: Option<i64>,
    pub r#type: String,
    pub name: String,
    pub blob_path: Option<String>,
    pub size: Option<i64>,
    pub hash: Option<String>,
    pub deleted_at: Option<i64>,
    pub delete_batch_id: Option<String>,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize)]
pub struct DriveNode {
    pub id: i64,
    pub parent_id: Option<i64>,
    pub r#type: String,
    pub name: String,
    pub size: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hash: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub deleted_at: Option<i64>,
    pub created_at: i64,
    pub updated_at: i64,
    pub ext: Option<String>,
    pub mime_type: Option<String>,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub path: String,
    #[serde(skip_serializing_if = "is_zero")]
    pub share_count: i64,
    #[serde(skip)]
    pub blob_path: Option<String>,
    #[serde(skip)]
    pub delete_batch_id: Option<String>,
}

fn is_zero(v: &i64) -> bool {
    *v == 0
}

impl DriveNode {
    pub fn from_row(row: DriveNodeRow) -> Self {
        let (ext, mime) = if row.r#type == "file" {
            let e = ext_of(&row.name);
            let m = mime_for_ext(&e);
            (Some(e), Some(m))
        } else {
            (None, None)
        };
        DriveNode {
            id: row.id,
            parent_id: row.parent_id,
            r#type: row.r#type,
            name: row.name,
            size: row.size,
            hash: row.hash,
            deleted_at: row.deleted_at,
            created_at: row.created_at,
            updated_at: row.updated_at,
            ext,
            mime_type: mime,
            path: String::new(),
            share_count: 0,
            blob_path: row.blob_path,
            delete_batch_id: row.delete_batch_id,
        }
    }
}

pub fn ext_of(name: &str) -> String {
    Path::new(name)
        .extension()
        .map(|s| format!(".{}", s.to_string_lossy().to_lowercase()))
        .unwrap_or_default()
}

pub fn mime_for_ext(ext: &str) -> String {
    let e = ext.trim_start_matches('.').to_lowercase();
    match e.as_str() {
        "html" | "htm" => "text/html; charset=utf-8".into(),
        "txt" => "text/plain; charset=utf-8".into(),
        "css" => "text/css; charset=utf-8".into(),
        "js" | "mjs" => "application/javascript".into(),
        "json" => "application/json".into(),
        "xml" => "application/xml".into(),
        "pdf" => "application/pdf".into(),
        "zip" => "application/zip".into(),
        "gz" | "tgz" => "application/gzip".into(),
        "tar" => "application/x-tar".into(),
        "rtf" => "application/rtf".into(),
        "csv" => "text/csv; charset=utf-8".into(),
        "md" | "markdown" => "text/markdown; charset=utf-8".into(),
        "png" => "image/png".into(),
        "jpg" | "jpeg" => "image/jpeg".into(),
        "gif" => "image/gif".into(),
        "webp" => "image/webp".into(),
        "bmp" => "image/bmp".into(),
        "tiff" | "tif" => "image/tiff".into(),
        "svg" => "image/svg+xml".into(),
        "ico" => "image/x-icon".into(),
        "mp3" => "audio/mpeg".into(),
        "wav" => "audio/wav".into(),
        "ogg" => "audio/ogg".into(),
        "flac" => "audio/flac".into(),
        "m4a" => "audio/mp4".into(),
        "mp4" => "video/mp4".into(),
        "webm" => "video/webm".into(),
        "mov" => "video/quicktime".into(),
        "mkv" => "video/x-matroska".into(),
        "avi" => "video/x-msvideo".into(),
        "doc" => "application/msword".into(),
        "docx" => "application/vnd.openxmlformats-officedocument.wordprocessingml.document".into(),
        "xls" => "application/vnd.ms-excel".into(),
        "xlsx" => "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet".into(),
        "ppt" => "application/vnd.ms-powerpoint".into(),
        "pptx" => {
            "application/vnd.openxmlformats-officedocument.presentationml.presentation".into()
        }
        "" => "application/octet-stream".into(),
        _ => "application/octet-stream".into(),
    }
}

#[derive(Debug, Clone, FromRow)]
pub struct DriveShareRow {
    pub id: i64,
    pub node_id: i64,
    pub token_hash: String,
    pub token_prefix: String,
    pub token: Option<String>,
    pub password_hash: Option<String>,
    pub expires_at: Option<i64>,
    pub created_at: i64,
}

#[derive(Debug, Clone, FromRow)]
pub struct DriveUpload {
    pub id: String,
    pub parent_id: Option<i64>,
    pub name: String,
    pub size: i64,
    pub chunk_size: i64,
    pub total_chunks: i64,
    pub received_mask: Vec<u8>,
    pub status: String,
    pub expires_at: i64,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Serialize)]
pub struct DriveBreadcrumb {
    pub id: i64,
    pub name: String,
}

// ----- request DTOs -----

fn deserialize_optional_i64<'de, D: serde::Deserializer<'de>>(
    de: D,
) -> Result<Option<i64>, D::Error> {
    use serde::de::Error;
    let s: Option<serde_json::Value> = Option::deserialize(de)?;
    Ok(match s {
        None | Some(serde_json::Value::Null) => None,
        Some(serde_json::Value::Number(n)) => n.as_i64(),
        Some(serde_json::Value::String(s)) => {
            if s.is_empty() {
                None
            } else {
                Some(s.parse().map_err(D::Error::custom)?)
            }
        }
        _ => None,
    })
}

#[derive(Debug, Deserialize, Default)]
pub struct DriveListQuery {
    #[serde(default, deserialize_with = "deserialize_optional_i64")]
    pub parent_id: Option<i64>,
    pub q: Option<String>,
    #[serde(default)]
    pub order_by: String,
    #[serde(default)]
    pub sort: String,
}

#[derive(Debug, Deserialize)]
pub struct DriveIdQuery {
    pub id: i64,
}

#[derive(Debug, Deserialize)]
pub struct DriveCreateFolderRequest {
    pub parent_id: Option<i64>,
    pub name: String,
}

#[derive(Debug, Deserialize)]
pub struct DriveRenameRequest {
    pub id: i64,
    pub name: String,
}

#[derive(Debug, Deserialize)]
pub struct DriveMoveRequest {
    pub ids: Vec<i64>,
    pub new_parent_id: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct DriveIdsRequest {
    pub ids: Vec<i64>,
}

#[derive(Debug, Deserialize)]
pub struct DriveRestoreRequest {
    pub id: i64,
}

#[derive(Debug, Deserialize)]
pub struct DriveUploadInitRequest {
    pub parent_id: Option<i64>,
    pub name: String,
    pub size: i64,
    #[serde(default)]
    pub chunk_size: i64,
}

#[derive(Debug, Serialize)]
pub struct DriveUploadInitResponse {
    pub upload_id: String,
    pub total_chunks: i64,
    pub chunk_size: i64,
    pub received_chunks: Vec<i64>,
}

#[derive(Debug, Serialize)]
pub struct DriveUploadStatusResponse {
    pub upload_id: String,
    pub total_chunks: i64,
    pub chunk_size: i64,
    pub size: i64,
    pub received_chunks: Vec<i64>,
    pub status: String,
}

#[derive(Debug, Deserialize)]
pub struct DriveUploadCompleteRequest {
    pub upload_id: String,
    #[serde(default)]
    pub on_collision: String,
}

#[derive(Debug, Deserialize)]
pub struct DriveShareCreateRequest {
    pub node_id: i64,
    pub password: Option<String>,
    pub expires_at: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct DriveShareRevokeRequest {
    pub share_id: i64,
}

#[derive(Debug, Deserialize, Default)]
pub struct DriveSharesAllQuery {
    #[serde(default)]
    pub include_expired: bool,
}

#[derive(Debug, Serialize)]
pub struct DriveShareDTO {
    pub id: i64,
    pub node_id: i64,
    pub has_password: bool,
    pub expires_at: Option<i64>,
    pub created_at: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct DriveSharedItemDTO {
    pub id: i64,
    pub node_id: i64,
    pub parent_id: Option<i64>,
    pub has_password: bool,
    pub expires_at: Option<i64>,
    pub created_at: i64,
    pub name: String,
    pub size: i64,
    pub path: String,
    pub node_type: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub mime_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    #[serde(skip)]
    pub token: Option<String>,
}
